File: 3.singleton/singleton.java
class Database {
    private static Database instance;
    private Database() {
        System.out.println("Conexión establecida.");
    }
    public static Database getInstance() {
        if (instance == null) {
            instance = new Database();
        }
        return instance;
    }
}
public class singleton {
    public static void main(String[] args) {
        Database db1 = Database.getInstance();
        Database db2 = Database.getInstance();
        System.out.println(db1 == db2); // true, misma instancia
    }
}

File: 3.singleton/singleton.ts
class Database {
    private static instance: Database;
    private constructor() {
        console.log("Conexión establecida.");
    }
    public static getInstance(): Database {
        if (!Database.instance) {
            Database.instance = new Database();
        }
        return Database.instance;
    }
}
const db1 = Database.getInstance();
const db2 = Database.getInstance();
console.log(db1 === db2); // true, misma instancia